Engine/DataFeeds/Enumerators/SynchronizingBaseDataEnumerator.cs

```diff
@@ -0,0 +1,52 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using QuantConnect.Data;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace QuantConnect.Lean.Engine.DataFeeds.Enumerators
+{
+    /// <summary>
+    /// Represents an enumerator capable of synchronizing other base data enumerators in time.
+    /// This assumes that all enumerators have data time stamped in the same time zone
+    /// </summary>
+    public class SynchronizingBaseDataEnumerator : SynchronizingEnumerator<BaseData>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SynchronizingBaseDataEnumerator"/> class
+        /// </summary>
+        /// <param name="enumerators">The enumerators to be synchronized. NOTE: Assumes the same time zone for all data</param>
+        public SynchronizingBaseDataEnumerator(params IEnumerator<BaseData>[] enumerators)
+            : this((IEnumerable<IEnumerator>)enumerators)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SynchronizingBaseDataEnumerator"/> class
+        /// </summary>
+        /// <param name="enumerators">The enumerators to be synchronized. NOTE: Assumes the same time zone for all data</param>
+        public SynchronizingBaseDataEnumerator(IEnumerable<IEnumerator> enumerators) : base((IEnumerable<IEnumerator<BaseData>>)enumerators)
+        {
+        }
+
+        protected override DateTime GetInstanceTime<T>(T instance)
+        {
+            return (instance as BaseData).EndTime;
```

Comment:
Shouldn't need to be casting here, since ur specifying `T`, the problem is that `<T>` in `GetInstanceTime<T>` is actually saying it's a different `T`.
Should be
```csharp
        protected abstract DateTime GetInstanceTime(T instance);


        protected override DateTime GetInstanceTime(BaseData instance)
        {
            return instance.EndTime;
        }
```